temporal-sdk/src/main/java/io/temporal/internal/worker/EagerActivitySlotsReservation.java

```diff
@@ -66,34 +74,43 @@ public void applyToRequest(RespondWorkflowTaskCompletedRequest.Builder mutableRe
   public void handleResponse(RespondWorkflowTaskCompletedResponse serverResponse) {
     int activityTasksCount = serverResponse.getActivityTasksCount();
     Preconditions.checkArgument(
-        activityTasksCount <= this.outstandingReservationSlotsCount,
+        activityTasksCount <= this.reservedSlots.size(),
         "Unexpectedly received %s eager activities though we only requested %s",
         activityTasksCount,
-        this.outstandingReservationSlotsCount);
+        this.reservedSlots.size());
 
-    releaseSlots(this.outstandingReservationSlotsCount - activityTasksCount);
+    // Release any slots that we won't be using
+    releaseSlots(this.reservedSlots.size() - activityTasksCount);
```

Comment:
Since the slots are reserved based on the task queue don't we also need to keep track of that when using the slot and releasing the slot?

Reply:
Interesting point. I wouldn't say they're _based on_ so much as that information is provided when making the reservation (implementations may or may not care). It does make sense, of course, that we might also want to provide that information when releasing.

Arguably, users could stuff that in the permit data, but maybe we want to make it first-class. In that case, I'd probably just stick it as a field in the permit. However, you could also argue that implementations that want to balance across multiple queues maybe don't want to associate a permit with a specific queue, in which case it makes more sense on some `ReleaseContext` object to provide as another arg when releasing.

The latter maybe makes a bit more sense to me?

Reply:
Added to release context

Reply:
Sorry to clarify what I meant is your slot reservations are not fungible since they are tied to the task queue they are requested for. So you can't just blindly release first x slots you need to release the slots based on task queue.

Reply:
Yeah... I was hoping not to retain the property in the _built_ workflow options since having mutually exclusive stuff represented in the built type is less than good from a type design perspective. It also means the logic for the threadpool size has to get pushed out of the options, but, that's not so bad.

Reply:
Ah, yeah that makes sense. I don't know that it necessarily must work that way, we could only make the guarantee that we'll return permits since they all came from the same supplier regardless of queue, but we can do it fairly easily and it's probably helpful so I'll make that change.